gc/base/ParallelDispatcher.cpp

```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
 
 	startUpThreads();
 }
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+/**
+ * Contract the thread pool by shutting down threads in pool to obtain newThreadCount.
+ *
+ * This API assumes  CRIU is the only consumer (not tested for general use).
+ * The following conditions are assumed while contracting the thread pool:
+ *     1) API can only be called once during VM lifetime
+ *     2) NOT holding exclusive VM access
+ *     3) Dispatcher is idle (no task can be dispatched)
+ *     4) Can't externally enter shutdown from another source
+ */
+void
+MM_ParallelDispatcher::contractThreadPool(MM_EnvironmentBase *env, uintptr_t newThreadCount)
+{
+	Assert_MM_false(_workerThreadsReservedForGC);
+	Assert_MM_false(_inShutdown);
+
+	Assert_MM_true(_threadShutdownCount == (_threadCountAtGCInit - 1));
+	Assert_MM_true(_threadCountMaximum == _extensions->gcThreadCount);
+	Assert_MM_true(_threadCountMaximum == _threadCountAtGCInit);
+
+	uintptr_t preShutdownThreadCount = _extensions->gcThreadCount;
+
+	Trc_MM_ParallelDispatcher_criu_contractThreadPool_Entry(preShutdownThreadCount, newThreadCount);
+
+	/* Can't shutdown the main thread (Dispatcher didn't start it) */
+	if (0 == newThreadCount) {
+		newThreadCount = 1;
+	}
+
+	if (newThreadCount < _threadCountMaximum) {
+		Trc_MM_ParallelDispatcher_criu_contractThreadPool_Attempt();
+
+		omrthread_monitor_enter(_workerThreadMutex);
+
+		_inShutdown = true;
+
+		/* Clip the thread pool past newThreadCount */
+		for (uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
+			_statusTable[index] = worker_status_dying;
+		}
+
+		omrthread_monitor_notify_all(_workerThreadMutex);
+
+		omrthread_monitor_exit(_workerThreadMutex);
+
+		uintptr_t expectedThreadShutdownThread = newThreadCount - 1;
+
+		omrthread_monitor_enter(_dispatcherMonitor);
+		while (expectedThreadShutdownThread != _threadShutdownCount) {
+			omrthread_monitor_wait(_dispatcherMonitor);
+		}
+		omrthread_monitor_exit(_dispatcherMonitor);
+
+		/* Cleanup the dispatcher tables */
+		for(uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
+			Assert_MM_true(worker_status_dying == _statusTable[index]);
+			_statusTable[index] = worker_status_inactive;
+			_threadTable[index] = NULL;
+		}
+
+		Assert_MM_true(_threadShutdownCount == expectedThreadShutdownThread);
+
+		_extensions->gcThreadCount = _threadCount = _threadCountMaximum = newThreadCount;
+		_activeThreadCount = adjustThreadCount(_threadCount);
```

Comment:
it's probably simpler to just set this to _threadCountMaximum, too.
it will be adjusted to a meaningful value at restore time, when it's really to be used, anyway